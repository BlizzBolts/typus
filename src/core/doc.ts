import ts from "typescript";
import type { SourceFile } from "./sourceFile";

export interface SerialziedDoc {
  name?: string;
  documentation?: string;
  type?: string;
  tags?: { name: string; text: string }[];
  required?: boolean;
  members?: SerialziedDoc[];
  parameters?: SerialziedDoc[];
  children?: SerialziedDoc[];
  symbol?: ts.Symbol;
}

export interface DocPayload {
  name?: string;
  documentation?: string;
  type?: string;
  tags?: { name: string; text: string }[];
  required?: boolean;
  members?: Doc[];
  parameters?: Doc[];
  children?: Doc[];
  symbol?: ts.Symbol;
}

export class Doc {
  name?: string;
  documentation?: string;
  type?: string;
  tags?: { name: string; text: string }[];
  required?: boolean;
  members?: Doc[];
  parameters?: Doc[];
  children?: Doc[];
  symbol?: ts.Symbol;

  root?: SourceFile | null;

  constructor(param: DocPayload) {
    this.name = param.name || "";
    this.documentation = param.documentation || "";
    this.type = param.type || "";
    this.tags = param.tags || [];
    this.required = param.required || false;
    this.members = param.members || [];
    this.parameters = param.parameters || [];
    this.children = param.children || [];
    this.symbol = param.symbol || (null as unknown as ts.Symbol);

    this.root = null;
  }

  public setRoot(root: SourceFile) {
    this.root = root;
  }

  static serialize = (curr: Doc): SerialziedDoc => {
    return {
      name: curr.name,
      documentation: curr.documentation,
      type: curr.type,
      tags: curr.tags,
      required: curr.required,
      members: curr.members?.map(Doc.serialize),
      parameters: curr.parameters?.map(Doc.serialize),
      children: curr.children?.map(Doc.serialize),
    };
  };
}
