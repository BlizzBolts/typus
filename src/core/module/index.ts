export { ModuleGraph } from "./graph";
export { Module } from "./module";
export type { BaseModule } from "./types";
