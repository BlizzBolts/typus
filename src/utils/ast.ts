import ts from "typescript"

export function isNodeExported(node: ts.Node): boolean {
  return (
    (ts.getCombinedModifierFlags(node as ts.Declaration) &
      ts.ModifierFlags.Export) !==
      0 ||
    (!!node.parent && node.parent.kind === ts.SyntaxKind.SourceFile)
  )
}
