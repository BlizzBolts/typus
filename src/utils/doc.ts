import ts from "typescript"
import type { DocEntry } from "../types"

// export const getInterfaceDocuments = (
//   node: ts.Node,
//   checker: ts.TypeChecker,
// ) => {
//   const symbol = checker.getSymbolAtLocation(node.name)
//   if (!symbol) {
//     return
//   }

//   const main = serializeSymbol(symbol, checker, node)

//   main.fields = node.members
//     .map((m) => {
//       if (!m.name) return
//       const symbol = checker.getSymbolAtLocation(m.name)
//       if (!symbol) return
//       return serializeSymbol(symbol, checker, m)
//     })
//     .filter(Boolean) as DocEntry[]

//   return main
// }

export const serializeSymbol = (
  symbol: ts.Symbol,
  checker: ts.TypeChecker,
): DocEntry => {
  return {
    name: symbol?.getName(),
    documentation: ts.displayPartsToString(
      symbol?.getDocumentationComment(checker),
    ),
    symbolType: checker.getTypeOfSymbolAtLocation(
      symbol,
      symbol?.valueDeclaration!,
    ),
    type: checker.typeToString(
      checker.getTypeOfSymbolAtLocation(symbol, symbol?.valueDeclaration!),
    ),
    tags: symbol?.getJsDocTags(checker).map((o) => {
      return {
        name: o.name,
        text: ts.displayPartsToString(o.text),
      }
    }),
    children: [],
    members: [],
    parameters: [],
    // @ts-ignore
    required: symbol?.["questionToken"],
  }
}
