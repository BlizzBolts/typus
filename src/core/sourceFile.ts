import { Doc, SerialziedDoc } from "./doc";

export type FilePath = string;

interface SourceFilePayload {
  filePath: FilePath;
  docs?: Doc[];
}

export interface SerialzieSourceFile {
  filePath: FilePath;
  docs: SerialziedDoc[];
}

export class SourceFile {
  private filePath: FilePath;

  private docs?: Doc[];

  constructor(payload: SourceFilePayload) {
    const { filePath, docs = [] } = payload;
    this.filePath = filePath;
    this.docs = docs;
  }

  public setFilePath(filePath: FilePath): this {
    this.filePath = filePath;
    return this;
  }

  public getFilePath(): FilePath {
    return this.filePath;
  }

  public addDoc(doc: Doc) {
    this.docs!.push(doc);
    return this;
  }

  public static serialize(curr: SourceFile) {
    return {
      filePath: curr.filePath,
      docs: curr.docs?.map(Doc.serialize),
    };
  }
}
