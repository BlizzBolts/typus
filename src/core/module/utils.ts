import ts, { SymbolDisplayPart } from "typescript";
import { Module } from "./module";

const MODULE_MAPPER = new Map<number, Module>();

export const getModuleByHash = (hashId: number): Module | undefined => {
  return MODULE_MAPPER.get(hashId);
};

export const setModule = (hash: number, m: Module): void => {
  MODULE_MAPPER.set(hash, m);
};

export const generateModule = (
  node: ts.Node,
  type: ts.Type,
  checker: ts.TypeChecker
): Module => {
  // @ts-ignore
  const symbol = checker.getSymbolAtLocation(node) || node.symbol;
  return new Module({
    name: symbol?.getName(),
    documentation: ts.displayPartsToString(
      symbol?.getDocumentationComment(checker)
    ),
    type: checker.typeToString(type),
    _type: type,
    tags: symbol
      ?.getJsDocTags(checker)
      .map((o: { name: string; text: SymbolDisplayPart[] }) => {
        return {
          name: o.name,
          text: ts.displayPartsToString(o.text),
        };
      }),
    children: [],
    members: [],
    parameters: [],
    symbol: checker.getTypeAtLocation(node) as unknown as ts.Symbol,
    // required: symbol?.["questionToken"],
  });
};

export const generateModuleForTypeAliasFunction = () => {};
