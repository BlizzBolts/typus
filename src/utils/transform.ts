import { flatten } from "lodash";
import { Doc } from "../core/doc";
import { Module, ModuleGraph } from "../core/module";

let cacheDocs = new Map<Module, Doc>();

export const transformModuleGraph2Doc = (graph: ModuleGraph): Doc => {
  const { children } = graph;
  const rootDoc = new Doc({
    name: "root",
  });
  const modules = flatten(children);

  rootDoc.children = modules.map((o) => traverseModule2Doc(o, graph));
  cacheDocs.clear();
  return rootDoc;
};

export const traverseModule2Doc = (module: Module, graph: ModuleGraph): Doc => {
  const doc = new Doc();
  doc.name = module.name;
  doc.documentation = module.documentation;
  doc.type = module.type;
  doc.tags = module.tags;
  doc.symbol = module.symbol;
  doc.required = module.required;

  // block cycle reference
  const cacheDoc = cacheDocs.get(module);
  if (cacheDoc) {
    return doc;
  } else {
    cacheDocs.set(module, doc);
  }

  doc.members = module?.members?.map((o) => traverseModule2Doc(o, graph)) || [];
  doc.parameters =
    module?.parameters?.map((o) => traverseModule2Doc(o, graph)) || [];
  doc.children =
    module?.children?.map((o) => traverseModule2Doc(o, graph)) || [];

  return doc;
};
