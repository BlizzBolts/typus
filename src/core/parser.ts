import ts from "typescript";
import { generateModule } from "../core/module/utils";
import { ModuleGraph, Module } from "./module";

export const defaultCompilerOptions = {
  jsx: ts.JsxEmit.React,
  module: ts.ModuleKind.CommonJS,
  target: ts.ScriptTarget.Latest,
};

export class Parser {
  private compilerOptions: ts.CompilerOptions = defaultCompilerOptions;
  private filePaths: string[] = [];
  private program!: ts.Program;
  private typeChecker!: ts.TypeChecker;
  private moduleGraph: ModuleGraph = new ModuleGraph();

  constructor(
    filePaths?: string | string[],
    compilerOptions: Partial<ts.CompilerOptions> = defaultCompilerOptions
  ) {
    if (filePaths) {
      this.setup(filePaths, compilerOptions);
    }
  }

  static of(
    filePaths?: string | string[],
    compilerOptions: Partial<ts.CompilerOptions> = defaultCompilerOptions
  ) {
    return new Parser(filePaths, compilerOptions);
  }

  getTypeChecker() {
    return this.typeChecker;
  }

  getProgram() {
    return this.program;
  }

  getModuleGraph() {
    return this.moduleGraph;
  }

  setup(
    filePaths: string | string[],
    compilerOptions: Partial<ts.CompilerOptions> = defaultCompilerOptions
  ): this {
    this.compilerOptions = Object.assign(
      defaultCompilerOptions,
      compilerOptions
    );
    this.filePaths = Array.isArray(filePaths) ? filePaths : [filePaths];
    this.program = ts.createProgram(this.filePaths, this.compilerOptions);
    this.typeChecker = this.program.getTypeChecker();

    return this;
  }

  parse(): this {
    const validSourceFiles = this.program.getSourceFiles().filter((o) => {
      return !o.isDeclarationFile;
    });

    const moduleGraphs = validSourceFiles
      .map((o: ts.SourceFile) => {
        const result: Module[] = [];
        o.forEachChild((n: ts.Node): void => {
          const module = this.traverse(n);
          if (module) {
            result.push(module);
          }
        });
        return result;
      })
      .filter(Boolean) as ModuleGraph[][];

    this.moduleGraph.children = moduleGraphs;
    return this;
  }

  parseInterfaceDeclaration(node: ts.InterfaceDeclaration): Module {
    const type = this.typeChecker.getTypeAtLocation(node.name);
    const cacheModule = this.moduleGraph.getCacheBy(type);
    if (cacheModule) {
      return cacheModule;
    }
    const module = generateModule(node, type, this.typeChecker);
    this.moduleGraph.setCacheModule(type, module);
    if (Array.isArray(node.members)) {
      module.members = node.members
        .map((memberNode) => {
          return this.traverse(memberNode);
        })
        .filter(Boolean) as ModuleGraph[];
    }
    return module;
  }

  parseTypeAliasDeclaration(node: ts.TypeAliasDeclaration): Module {
    const type = this.typeChecker.getTypeAtLocation(node.name);
    const cacheModule = this.moduleGraph.getCacheBy(type);
    if (cacheModule) {
      return cacheModule;
    }
    if (ts.isTypeAliasDeclaration(node)) {
      console.log(1);
    }
    const module = generateModule(node, type, this.typeChecker);
    this.moduleGraph.setCacheModule(type, module);
    return module;
  }

  parseMethodSignature(node: ts.MethodSignature): Module {
    const type = this.typeChecker.getTypeAtLocation(node);
    const cacheModule = this.moduleGraph.getCacheBy(type);
    if (cacheModule) {
      return cacheModule;
    }
    const module = generateModule(node, type, this.typeChecker);

    if (Array.isArray(node.parameters)) {
      module.parameters = node.parameters
        .map((parameter) => this.traverse(parameter))
        .filter(Boolean) as ModuleGraph[];
    }
    return module;
  }

  parseParameter = (node: ts.ParameterDeclaration): Module => {
    const type = this.typeChecker.getTypeAtLocation(node);
    const cacheModule = this.moduleGraph.getCacheBy(type);
    if (cacheModule) {
      return cacheModule;
    }
    const module = generateModule(node.name, type, this.typeChecker);
    return module;
  };

  parsePropertySignature = (node: ts.PropertySignature) => {
    const type = this.typeChecker.getTypeAtLocation(node.name!);
    const cacheModule = this.moduleGraph.getCacheBy(type);
    if (cacheModule) {
      return cacheModule;
    }
    const module = generateModule(node.name, type, this.typeChecker);
    if (node.type && ts.isFunctionTypeNode(node.type)) {
      module.parameters = node.type.parameters
        .map((parameter) => this.traverse(parameter))
        .filter(Boolean) as ModuleGraph[];
    }
    return module;

    // if (node.type && ts.isTypeReferenceNode(node.type)) {
    //   node.type.parameters.forEach((parameter) =>
    //     this.traverse(parameter, doc)
    //   );
    // }
  };

  // parseFunction(node: ts.FunctionDeclaration, parent: Doc) {
  //   const fnType = this.typeChecker.getTypeAtLocation(node);
  //   const Doc = generateModule(node, fnType.getSymbol()!, this.typeChecker);
  //   parent.children?.push(Doc);
  //   this.cache.set(fnType, Doc);

  //   if (Array.isArray(node.parameters)) {
  //     node.parameters.forEach((parameter) => this.traverse(parameter, Doc));
  //   }
  // }

  // parseInterfaceMember = (node: ts.TypeElement, parent: Doc) => {
  //   const memberType = this.typeChecker.getTypeAtLocation(node!.name!);

  //   if (this.cache.has(memberType)) {
  //     parent.members?.push(this.cache.get(memberType)!);
  //   } else {
  //     const symbol = this.typeChecker.getSymbolAtLocation(node.name!);
  //     parent.members?.push(generateModule(node, symbol!, this.typeChecker));
  //   }
  // };

  traverse(node: ts.Node): Module | null {
    // if (this.cache.has(this.typeChecker.getTypeAtLocation(node))) {
    // }

    if (ts.isInterfaceDeclaration(node)) {
      console.log("isInterfaceDeclaration!");
      return this.parseInterfaceDeclaration(node);
    }

    if (ts.isMethodSignature(node)) {
      console.log("isMethodSignature!");
      return this.parseMethodSignature(node);
    }

    if (ts.isPropertySignature(node)) {
      console.log("isPropertySignature!");
      return this.parsePropertySignature(node);
    }

    if (ts.isParameter(node)) {
      console.log("isParameter!");
      return this.parseParameter(node);
    }

    // if (ts.isVariableDeclaration(node)) {
    //   console.log("isVariableDeclaration!");
    //   this.parseIndexSignatureDeclaration(node, parent)
    //   return;
    // }

    // if (ts.isFunctionDeclaration(node)) {
    //   console.log("isFunctionDeclaration!");
    //   this.parseFunction(node, parent);
    //   return;
    // }

    // if (ts.isClassDeclaration(node)) {
    //   console.log("isClassDeclaration");
    //   return;
    // }

    if (ts.isTypeAliasDeclaration(node)) {
      console.log("isTypeAliasDeclaration");
      return this.parseTypeAliasDeclaration(node);
    }

    // if (ts.isEnumDeclaration(node)) {
    //   console.log("isEnumDeclaration");
    //   return;
    // }
    // if (ts.isModuleDeclaration(node)) {
    //   console.log("isModuleDeclaration");
    //   return;
    // }

    // if (ts.isTypeElement(node)) {
    //   console.log("isTypeElement!");
    //   this.parseInterfaceMember(node, parent);
    //   return;
    // }

    return null;
  }
}
