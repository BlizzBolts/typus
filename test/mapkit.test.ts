import { it, expect, suite } from "vitest";
import path from "path";
import { Parser } from "../src/core/parser";
import { Doc } from "../src/core/doc";
import { transformModuleGraph2Doc } from "../src/utils/transform";

suite("parse mapkit", () => {
  it("should parse mapkit", () => {
    const parser = Parser.of()
      .setup(path.resolve(__dirname, "./manifest/mapkit.ts"))
      .parse();
    const docCache = transformModuleGraph2Doc(parser.getModuleGraph());
    expect(Doc.serialize(docCache)).toMatchSnapshot();
  });
});
