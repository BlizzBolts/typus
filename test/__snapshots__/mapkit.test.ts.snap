// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse mapkit > should parse mapkit 1`] = `
{
  "children": [
    {
      "children": [],
      "documentation": "",
      "members": [
        {
          "children": [],
          "documentation": "上报日志",
          "members": [],
          "name": "log",
          "parameters": [
            {
              "children": [],
              "documentation": "",
              "members": [],
              "name": "payload",
              "parameters": [],
              "required": false,
              "tags": [],
              "type": "{ content: string; v: number; }",
            },
          ],
          "required": false,
          "tags": [
            {
              "name": "date",
              "text": "7/4/2023 - 4:43:53 PM",
            },
            {
              "name": "type",
              "text": "{(payload: { content: string, v: number }) => Promise<void>}",
            },
          ],
          "type": "(payload: { content: string; v: number; }) => Promise<void>",
        },
      ],
      "name": "ActionMeta",
      "parameters": [],
      "required": false,
      "tags": [],
      "type": "any",
    },
    {
      "children": [],
      "documentation": "",
      "members": [
        {
          "children": [],
          "documentation": "",
          "members": [
            {
              "children": [],
              "documentation": "上报日志",
              "members": [],
              "name": "log",
              "parameters": [
                {
                  "children": [],
                  "documentation": "",
                  "members": [],
                  "name": "payload",
                  "parameters": [],
                  "required": false,
                  "tags": [],
                  "type": "{ content: string; v: number; }",
                },
              ],
              "required": false,
              "tags": [
                {
                  "name": "date",
                  "text": "7/4/2023 - 4:43:53 PM",
                },
                {
                  "name": "type",
                  "text": "{(payload: { content: string, v: number }) => Promise<void>}",
                },
              ],
              "type": "(payload: { content: string; v: number; }) => Promise<void>",
            },
          ],
          "name": "ActionMeta",
          "parameters": [],
          "required": false,
          "tags": [],
          "type": "any",
        },
      ],
      "name": "Action",
      "parameters": [],
      "required": false,
      "tags": [],
      "type": "any",
    },
    {
      "children": [],
      "documentation": "",
      "members": [
        {
          "children": [],
          "documentation": "",
          "members": [
            {
              "children": [],
              "documentation": "",
              "members": [
                {
                  "children": [],
                  "documentation": "上报日志",
                  "members": [],
                  "name": "log",
                  "parameters": [
                    {
                      "children": [],
                      "documentation": "",
                      "members": [],
                      "name": "payload",
                      "parameters": [],
                      "required": false,
                      "tags": [],
                      "type": "{ content: string; v: number; }",
                    },
                  ],
                  "required": false,
                  "tags": [
                    {
                      "name": "date",
                      "text": "7/4/2023 - 4:43:53 PM",
                    },
                    {
                      "name": "type",
                      "text": "{(payload: { content: string, v: number }) => Promise<void>}",
                    },
                  ],
                  "type": "(payload: { content: string; v: number; }) => Promise<void>",
                },
              ],
              "name": "ActionMeta",
              "parameters": [],
              "required": false,
              "tags": [],
              "type": "any",
            },
          ],
          "name": "Action",
          "parameters": [],
          "required": false,
          "tags": [],
          "type": "any",
        },
      ],
      "name": "MapkitType",
      "parameters": [],
      "required": false,
      "tags": [],
      "type": "any",
    },
  ],
  "documentation": "",
  "members": [],
  "name": "root",
  "parameters": [],
  "required": false,
  "tags": [],
  "type": "",
}
`;
